### PR TITLE
Start on experimental esbuild support

### LINKED ISSUE #588: Investigate esbuild for `skuba build`
This presents a few opportunities while still preserving partial compatibility with our existing `tsconfig.json`-based builds with `tsc`:

- Bundle code to improve cold start time, particularly in AWS Lambda
- Extend the build process with plugins, including build-time module alias resolution
- In-process execution, as it supports a convenient Node.js API while `tsc` is CLI-only
- Faster builds, as it skips type checking which we already handle on lint

### PR DESCRIPTION
**Pull request by 72636c**
This is very MVPR but I suppose it closes #588, at least from an investigative standpoint.

### CODE REVIEW

**Comment by samchungy** [comment]
🕺 💃 :D :D

**Comment by 72636c** [comment]
This has been collecting dust for too long and I think it would be more productive to chip away at the integration in smaller PRs against the default branch. It seems I've gotten far enough to get skuba to build itself, though there are a tonne of missing features and we probably need to clean up the error/warning output.

You're way more experienced with esbuild @samchungy and we'll likely lean on you heavily here 🙂.

File: src/global.d.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// evanw/esbuild#2388
+
+declare global {
+  namespace WebAssembly {
+    interface Module {}
+  }
+}
```
Comment (72636c) [inline src/global.d.ts:7]: Oops, forgot one thing: we need to figure out a way to include declaration in skuba's bundle and for TypeScript to pick it up in consuming repos, else you'll get compiler errors when type checking esbuild itself unless you have `DOM` types or `skipLibCheck: true`.

**Comment by samchungy** [comment]
> This has been collecting dust for too long and I think it would be more productive to chip away at the integration in smaller PRs against the default branch. It seems I've gotten far enough to get skuba to build itself, though there are a tonne of missing features and we probably need to clean up the error/warning output.
> 
> You're way more experienced with esbuild @samchungy and we'll likely lean on you heavily here 🙂.

Tbh though I haven't used esbuild without `bundle` so I'd need a bit of time to 👀 at everything

File: src/cli/build/esbuild.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    format: compilerOptions.module === ModuleKind.CommonJS ? 'cjs' : undefined,
+    outdir: compilerOptions.outDir,
+    logLevel: debug ? 'debug' : 'info',
+    logLimit: 0,
+    platform:
+      compilerOptions.moduleResolution === ModuleResolutionKind.NodeJs
+        ? 'node'
+        : undefined,
+    plugins: [
+      tsconfigPaths({
```
Comment (samchungy) [inline src/cli/build/esbuild.ts:104]: Interesting - is the use of this plugin because of https://github.com/evanw/esbuild/issues/394?

**Review by samchungy — APPROVED** [review]

File: src/cli/build/esbuild.ts (hunk 2 above)
Comment (72636c) [reply src/cli/build/esbuild.ts:104]: Yep, spot on; I'll include this as a comment. Not super enthused by the v0 plugin but it's the simplest path for now; if we end up building our own plugins for more complex scenarios I reckon we could vendor/improve on this.

File: src/global.d.ts (hunk 1 above)
Comment (72636c) [reply src/global.d.ts]: I think 1494857 will do it, but this upcoming release probably necessitates a beta or two to ensure everything comes together correctly.

**Pull request merged**